### PR TITLE
docs: Formalize our design docs process a bit.

### PR DESCRIPTION
**Pull request by PhilipMetzger**
This adds the basic outline of _when_ a Design Doc should be written. See the next
commit in the stack for the blueprint. 

By adding this we hopefully can prevent unnecessary churn from new and longtime contributors, 
when they want to add a major feature or rewrite a core part of Jujutsu. The text is written
as a guideline, not a rule.

### CODE REVIEW

**Review by fowles — APPROVED** [review]

File: docs/design_docs.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+Jujutsu uses Design Docs to drive technical decisions on large projects and it
+is the place to discuss your proposed design or new component. It is a very 
+thorough process, in which all concerns must be addressed. It shares some
+similarities with [Rust RFCs] but mostly addresses _technical_ problems and 
+gauges the technical and social concerns of all stakeholders.
+
+So if you want to start building the native backend or the server component for
+Jujutsu, you'll need to go through this process. It also is good if you're 
+already familiar with the codebase, to not stall the implementation.
```
Comment (martinvonz) [inline docs/design_docs.md:11]: Maybe something like "It's also good if you're already familiar with the codebase" => "It's good to go through the process even if you're already familiar with the codebase". I wasn't sure if you meant that or if you meant that one should get familiar with the codebase first ("It's good if you're familiar with the codebase before starting work").

File: docs/contributing.md [hunk 2]
```diff
@@ -64,6 +64,12 @@ your company's interests, do feel free to approve it.
 This project follows [Google's Open Source Community
 Guidelines](https://opensource.google/conduct/).
 
+## Contributing large patches
+
+While we appreciate that you took the time to design/redesign or rework an 
+existing component, we cannot accept it as easily as that. As it requires a
+architecture review from multiple stakeholders, which we do with 
+[Design Docs](design_docs.md), see the [process here](design_docs.md#process).
```
Comment (martinvonz) [inline docs/contributing.md:72]: nit: This sounds like it's written for someone who has already made the mistake of sending a patch for a large feature without asking about it first. I think it sounds friendlier to say something like "Before sending a PR for a large change, please ..."

**Review by martinvonz — COMMENTED** [review]
Thanks!

File: docs/design_doc_blueprint.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+## State of the Feature as of `$VERSION` (optional)
+
+The state of the feature you want to improve and where it currently falls 
+short. If there's nothing to compare to, leave it out.
+
+## Goals and non-goals
+
+Direct goals of the project and features deemed not worth pursuing.
+
+## Overview
```
Comment (matts1) [inline docs/design_doc_blueprint.md:23]: Might also be useful to add the following:
* What assumed knowledge is there
  * Is there any place you can read up on any assumed knowledge that might be required to fully understand this document?
* Related work
  * Eg. Have any other VCS's addressed this problem? If so:
    * How did they do it
    * Pros / cons of the way they did it, would it work for jj?

**Review by matts1 — APPROVED** [review]

File: docs/design_docs.md (hunk 1 above)
Comment (PhilipMetzger) [reply docs/design_docs.md:11]: I intended to say the second thing, as people unfamiliar with the codebase shouldn't propose large changes.

File: docs/contributing.md (hunk 2 above)
Comment (PhilipMetzger) [reply docs/contributing.md]: rewritten to be friendlier

File: docs/design_doc_blueprint.md (hunk 3 above)
Comment (PhilipMetzger) [reply docs/design_doc_blueprint.md:29]: I added a prior and related work section.

File: docs/design_docs.md (hunk 1 above)
Comment (PhilipMetzger) [reply docs/design_docs.md]: now rewritten as that

Comment (martinvonz) [reply docs/design_docs.md]: I see. I don't see how that's very related to the previous sentence in the paragraph. And isn't it true whether or not they're writing a design doc? That is, it seems like it belongs more in `contributing.md` than here, but it also seems obvious enough that I don't think we need to even mention it.

Comment (PhilipMetzger) [reply docs/design_docs.md]: Ok, should I remove or rewrite it to your first proposed variant?

Comment (martinvonz) [reply docs/design_docs.md]: Just remove it IMO, but feel free to say you disagree, of course

Comment (PhilipMetzger) [reply docs/design_docs.md]: removed

**Pull request merged**